pep-lint.py

```diff
@@ -0,0 +1,427 @@
+            yield from _date(line_num, post_date, "Post-History")
+            yield from _thread(line_num, post_url, "Post-History")
+
+
+def _validate_resolution(line_num, line):
+    """'Resolution' must be a direct thread/message URL"""
+
+    yield from _thread(line_num, line, "Resolution", allow_message=True)
+
+
+########################
+#  Validation Helpers  #
+########################
+
+def _pep_num(line_num, pep_number, prefix):
+    if pep_number.startswith("0") and pep_number != "0":
+        yield line_num, f"{prefix} must not contain leading zeros: {pep_number}"
+    if not _is_digits(pep_number):
+        yield line_num, f"{prefix} must be numeric: {pep_number}"
+    elif not 0 <= int(pep_number) <= 9999:
+        yield line_num, f"{prefix} must be between 0 and 9999: {pep_number}"
+
+
+def _is_digits(string):
+    """Match a string of ASCII digits ([0-9]+)."""
+    return string.isascii() and string.isdigit()
+
+
+def _email(line_num, author_email, prefix):
+    if EMAIL_PATTERN.match(author_email.strip()) is None:
+        yield line_num, f"{prefix} must be a list of 'Name <email@example.com>' entries: {author_email}"
+
+
+def _thread(line_num, url, prefix, allow_message=False, discussions_to=False):
+    msg = f"{prefix} must be a valid thread URL"
+
+    if not url.startswith("https://"):
+        if not discussions_to:
+            yield line_num, msg
+        return
+
+    if url.startswith("https://discuss.python.org/t/"):
+        # Discussions-To links must be the thread itself, not a post
+        pattern = DISCOURSE_THREAD_PATTERN if discussions_to else DISCOURSE_POST_PATTERN
+
+        remainder = url.removeprefix("https://discuss.python.org/t/").removesuffix("/")
+        if pattern.fullmatch(remainder) is not None:
+            return
+
+    if url.startswith("https://mail.python.org/pipermail/"):
+        remainder = url.removeprefix("https://mail.python.org/pipermail/")
+        if MAILMAN_2_PATTERN.fullmatch(remainder) is not None:
+            return
+
+    if url.startswith("https://mail.python.org/archives/list/"):
+        remainder = url.removeprefix("https://mail.python.org/archives/list/")
+        if allow_message and MAILMAN_3_MESSAGE_PATTERN.fullmatch(remainder) is not None:
+            return
+        if MAILMAN_3_THREAD_PATTERN.fullmatch(remainder) is not None:
+            return
+
+    yield line_num, msg
+
+
+def _date(line_num, date_str, prefix):
+    try:
+        parsed_date = dt.datetime.strptime(date_str, "%d-%b-%Y")
+    except ValueError:
+        yield line_num, f"{prefix} must be a 'DD-mmm-YYYY' date: {date_str}"
+        return
+
+    if parsed_date.year < 1990:
+        yield line_num, f"{prefix} must not be before Python was invented: {date_str}"
+    if parsed_date > (dt.datetime.now() + dt.timedelta(days=14)):
+        yield line_num, f"{prefix} must not be in the future: {date_str}"
+
+
+if __name__ == "__main__":
+    raise SystemExit(check())
```

Comment:
```suggestion
    raise sys.exit(check())
```

Is more idiomatic and far more common, especially at the top level (inside a non-main function I'd tend to `raise SystemExit`, in the relatively uncommon cases where that made sense).

For reference, I see [123 000 usages of `sys.exit`](https://grep.app/search?q=sys.exit%28&case=true&words=true&filter[lang][0]=Python) vs [5 100 of `raise SystemExit`](https://grep.app/search?q=raise%20SystemExit%28&case=true&words=true&filter[lang][0]=Python) in Python files indexed by grep.app, a nearly 25 to 1 ratio. Searching for each inside the `if __name__ == "__main__"` block, I see [9700 instances for `sys.exit`](https://grep.app/search?q=if%20__name__%20%3D%3D%20%5B%27%22%5D__main__%5B%22%27%5D%3A%5Cn%20%20%28%20%20%29%3F%28%20%20%20%20%29%3Fsys%5C.exit%5C%28&regexp=true&case=true&filter[lang][0]=Python) vs [635 instances of `raise SystemExit`](https://grep.app/search?q=if%20__name__%20%3D%3D%20%5B%27%22%5D__main__%5B%22%27%5D%3A%5Cn%20%20%28%20%20%29%3F%28%20%20%20%20%29%3Fraise%20SystemExit%5C%28&regexp=true&case=true&filter[lang][0]=Python), an over 15 to 1 ratio.

Reply:
(Fixing typo in the suggestion) 

```suggestion
    sys.exit(check())
```

---

I've seen the argument that `sys.exit`hides the detail that an exception is raised, and the extra import could be inconvenient (but we need it for other things). 

https://adamj.eu/tech/2021/10/10/the-many-ways-to-exit-in-python/

I don't really mind which we use.

Reply:
I prefer ``raise SystemExit``, the ``sys.exit`` docs also say that it just raises ``SystemExit`` internally so this just removes the indirection.